src/ProjectSerializer.cpp

```diff
@@ -89,115 +94,261 @@ XO("This recovery file was saved by Audacity 2.3.0 or before.\n"
-   using Int = std::int32_t;
+   template <typename T> void WriteAttr(const std::string_view& name, T value)
+   {
+      assert(mInTag);
 
-   using Long = std::int32_t;   // To save long values
-   using ULong = std::uint32_t; // To save size_t values
+      if (!mInTag)
+         return;
 
-   using LongLong = std::int64_t;
+      mAttributes.emplace_back(name, XMLAttributeValueView(value));
+   }
 
-   // Detect this computer's endianness
-   bool IsLittleEndian()
+   void WriteData(std::string value)
    {
-      const std::uint32_t x = 1u;
-      return
-         static_cast<const unsigned char*>(static_cast<const void*>(&x))[0];
-      // We will assume the same for other widths!
+      if (mInTag)
+         EmitStartTag();
+
+      if (XMLTagHandler* const handler = mHandlers.back())
+         handler->HandleXMLContent(CacheString(std::move(value)));
    }
-   // In C++20 this could be
-   // constexpr bool IsLittleEndian = (std::endian::native == std::endian::little);
-   // static_assert( IsLittleEndian || (std::endian::native == std::endian::big),
-   //    "Oh no!  I'm mixed-endian!" );
-
-   // Functions that can read and write native integer types to a canonicalized
-   // little-endian file format.  (We don't bother to do the same for floating
-   // point numbers.)
-
-   // Write native little-endian to little-endian file format
-   template< typename Number >
-   void WriteLittleEndian( MemoryStream &out, Number value )
+
+   void WriteRaw(std::string value)
    {
-      out.AppendData( &value, sizeof(value) );
    }
 
-   // Write native big-endian to little-endian file format
-   template< typename Number >
-   void WriteBigEndian( MemoryStream &out, Number value )
+   bool Finalize()
    {
-      auto begin = static_cast<unsigned char*>( static_cast<void*>( &value ) );
-      std::reverse( begin, begin + sizeof( value ) );
-      out.AppendData( &value, sizeof(value) );
+      if (mInTag)
+      {
+         EmitStartTag();
+         EndTag(mCurrentTagName);
+      }
+
+      return mBaseHandler != nullptr;
    }
 
-   // Read little-endian file format to native little-endian
-   template< typename Number >
-   Number ReadLittleEndian( wxMemoryInputStream &in )
+private:
+   void EmitStartTag()
    {
-      Number result;
-      in.Read( &result, sizeof(result) );
-      return result;
+      if (mHandlers.empty())
+      {
+         mHandlers.push_back(mBaseHandler);
+      }
+      else
+      {
+         if (XMLTagHandler* const handler = mHandlers.back())
+            mHandlers.push_back(handler->HandleXMLChild(mCurrentTagName));
+         else
+            mHandlers.push_back(NULL);
+      }
+
+      if (XMLTagHandler*& handler = mHandlers.back())
+      {
+         if (!handler->HandleXMLTag(mCurrentTagName, mAttributes))
+         {
+            handler = nullptr;
+
+            if (mHandlers.size() == 1)
+               mBaseHandler = nullptr;
+         }
+      }
+
+      mStringsCache.clear();
```

Comment:
mStringsCache is a deque, I assume because you need non-relocation of the strings that you cache in it (so that you can make string_views later).

But could repeated clearing of the deque and then re-allocation in it doing some wasteful reallocations?  Are all pages freed when you clear()?  Standard says std::vector::clear() does not change capacity, but I don't see the same stipulated for deque.

Maybe you can keep one dummy string in it, and resize to 1 instead of clearing, if it matters.

Reply:
You really don't like deque ;-)

Usually, the problem with deque is considered to be different: the memory costs just for one item are very high because deque allocates in large chunks. std::sting is usually around 24-32 bytes (why, libc++?. No, really, why 32 bytes?). On Clang deque allocates no less the 4k, on GCC - 512 bytes, which surely fits quite a bit of strings in. Even horrible MSVC implementation is not a performance concern here. We have worse problems to be solved first, for example, `Sequence::HandleXMLTag` is currently the #1 issue, because it never estimates the resulting blocks count, so `mBlock.push_back(wb);` is consuming most of the time right now. Still, we have a 40x speed up, and that problem can be easily addressed in a separate, much smaller, and easier to review PR.

In case we ever end up having deque to be a performance problem here - the fix will be as easy as a custom allocator. And trust me - you have to write them a lot when you are fighting for the performance :-)

Reply:
Custom allocator.  I was wondering if you would find use for `std::pmr::string` and some other allocator for the character data.